src/java.desktop/windows/native/libawt/windows/awt_PrintJob.cpp

```diff
@@ -605,22 +621,38 @@ Java_sun_awt_windows_WPageDialogPeer__1show(JNIEnv *env, jobject peer)
      */
     pageFormatToSetup(env, self, page, &setup, AwtPrintControl::getPrintDC(env, self));
     if (env->ExceptionCheck()) {
-        CLEANUP_SHOW;
-        return doIt;
+        return JNI_FALSE;
     }
 
     setup.lpfnPageSetupHook = reinterpret_cast<LPPAGESETUPHOOK>(pageDlgHook);
     setup.Flags = PSD_ENABLEPAGESETUPHOOK | PSD_MARGINS;
 
     AwtDialog::CheckInstallModalHook();
 
+    // Undo modal hooks and save printer parameters
+    auto postCleanup = [&] {
+        AwtDialog::CheckUninstallModalHook();
+
+        AwtDialog::ModalActivateNextWindow(NULL, target, peer);
+
+        HGLOBAL oldG = AwtPrintControl::getPrintHDMode(env, self);
+        if (setup.hDevMode != oldG) {
+           AwtPrintControl::setPrintHDMode(env, self, setup.hDevMode);
```

Comment:
what does setPrintHDMode and setPrintHDName do? do they need to be called as part of cleanup even if there is an exception and we return JNI_FALSE?

Reply:
i think this code preserves the printer settings after the dialog closes (and only on success before this fix), so is it possible in the case the printer is removed and causes an exception, we shouldn't preserve these settings?